dotnet/Razorvine.Pyrolite/Pyrolite/Pickle/Unpickler.cs

```diff
@@ -93,24 +70,45 @@ public class Unpickler : IDisposable {
 	 * @return the reconstituted object hierarchy specified in the file.
 	 */
 	public object load(Stream stream) {
-		input = stream;
-		stack = new UnpickleStack();
-		while (true) {
-			byte key = PickleUtils.readbyte(input);
-			object value = dispatch(key);
-			if(value != NO_RETURN_VALUE)
-				return value;
-		}
-	}
+        stack = new UnpickleStack();
+        var unpickler = new Unpickler<StreamReader>(new StreamReader(stream), memo, stack, this);
+        return unpickler.Load();
+    }
 
-	/**
-	 * Read a pickled object representation from the given pickle data bytes.
-	 * 
-	 * @return the reconstituted object hierarchy specified in the file.
-	 */
-	public object loads(byte[] pickledata) {
-		return load(new MemoryStream(pickledata));
-	}
+    /// <summary>
+    /// Read a pickled object representation from the given pickle data bytes.
+    /// </summary>
+    /// <param name="pickledata">Serialized pickle data.</param>
+    /// <returns>the reconstituted object hierarchy specified in the memory buffer.</returns>
+    public object loads(byte[] pickledata) {
+        stack = new UnpickleStack();
+        var unpickler = new Unpickler<ArrayReader>(new ArrayReader(pickledata), memo, stack, this);
+        return unpickler.Load();
+    }
+
+    /// <summary>
+    /// Read a pickled object representation from the given pickle data bytes.
+    /// </summary>
+    /// <param name="pickledata">Serialized pickle data.</param>
+    /// <param name="stackCapacity">Initial capacity of the UnpickleStack.</param>
+    /// <returns>the reconstituted object hierarchy specified in the memory buffer.</returns>
+    public object loads(byte[] pickledata, int stackCapacity) {
+        stack = new UnpickleStack(stackCapacity);
+        var unpickler = new Unpickler<ArrayReader>(new ArrayReader(pickledata), memo, stack, this);
+        return unpickler.Load();
+    }
+
+    /// <summary>
+    /// Read a pickled object representation from the given pickle data memory buffer.
+    /// </summary>
+    /// <param name="pickledata">Serialized pickle data.</param>
+    /// <param name="stackCapacity">Optional parameter that suggests the initial capacity of stack. The default value is 4.</param>
+    /// <returns>the reconstituted object hierarchy specified in the memory buffer.</returns>
+    public object loads(ReadOnlyMemory<byte> pickledata, int stackCapacity = UnpickleStack.DefaultCapacity) {
+        stack = new UnpickleStack(stackCapacity);
+        var unpickler = new Unpickler<ReadOnlyMemoryReader>(new ReadOnlyMemoryReader(pickledata), memo, stack, this);
```

Comment:
If you really want to eek out every last bit of perf, and your measurements show that the ArrayReader is faster than the ReadOnlyMemoryReader, how about:
```C#
var unpickler = MemoryMarshal.TryGetArray(pickleddata, out ArraySegment array) ?
    new Unpickler<ArrayReader>(new ArrayReader(array.Array, array.Offset), memo, stack, this) :
    new Unpickler<ReadOnlyMemoryReader>(new ReadOnlyMemoryReader(pickledata), memo, stack, this);
```
or something like that?

Reply:
good idea! `MemoryMarshal.TryGetArray` is so hidden that I forgot about it